File: security.py
"""Security module — Bash allowlist and filesystem confinement.

Provides a `can_use_tool` callback factory for the Claude Code SDK that
enforces command allowlists and workspace-scoped file access.
"""

from __future__ import annotations

import os
import re
import shlex

from claude_code_sdk import (
    PermissionResult,
    PermissionResultAllow,
    PermissionResultDeny,
    ToolPermissionContext,
)


_SEGMENT_SPLIT_RE = re.compile(r"&&|\|\||;|\||\n")


def _split_segments(command: str) -> list[str]:
    """Split a shell command into top-level segments.

    Splits on ``&&``, ``||``, ``;``, ``|``, and newlines. Shell quoting is not
    considered — a segment split inside quotes would mis-detect but still fails
    safely because each half is re-validated.
    """
    return [seg.strip() for seg in _SEGMENT_SPLIT_RE.split(command) if seg.strip()]


def _segment_binary(segment: str) -> str | None:
    """Extract the effective binary from a single segment.

    Handles:
      - Simple commands: ``npm install``
      - Env-prefixed commands: ``NODE_ENV=production npm start``
    """
    try:
        tokens = shlex.split(segment)
    except ValueError:
        return None
    if not tokens:
        return None
    for token in tokens:
        if "=" in token and not token.startswith("-"):
            continue
        return os.path.basename(token)
    return None


def _extract_paths(segment: str) -> list[str]:
    """Return non-flag tokens from a segment (candidate filesystem paths)."""
    try:
        tokens = shlex.split(segment)
    except ValueError:
        return []
    return [t for t in tokens[1:] if not t.startswith("-") and "=" not in t]


def _is_inside(path: str, root: str) -> bool:
    """Return *True* iff ``path`` resolves inside ``root``."""
    real = os.path.realpath(path)
    return real == root or real.startswith(root + os.sep)


def create_permission_handler(
    config: dict,
    workspace_dir: str,
) -> "CanUseToolCallback":
    """Return an async ``can_use_tool`` callback bound to *config* and *workspace_dir*.

    The callback enforces:
    1. **Bash command allowlist** — every segment of a chained command must use
       a binary listed in ``config["security"]["bash_allowlist"]``.
    2. **Bash deny patterns** — commands matching any pattern in
       ``config["security"]["bash_denylist"]`` are blocked.
    3. **Bash path confinement** — any ``cd <path>`` must target a directory
       inside *workspace_dir*, and no segment may reference an absolute path
       under another workspace sibling (e.g. a different project).
    4. **Filesystem confinement for Write / Edit** — ``file_path`` must resolve
       inside *workspace_dir*.
    """
    allowlist: set[str] = set(config["security"]["bash_allowlist"])
    deny_patterns: list[re.Pattern[str]] = [
        re.compile(re.escape(p)) for p in config["security"]["bash_denylist"]
    ]
    abs_workspace = os.path.realpath(workspace_dir)
    # Parent workspace root (e.g. ``.../workspace/``) — used to detect attempts
    # to reach sibling projects. If the current workspace is already the top of
    # the filesystem, this degrades gracefully to the workspace itself.
    siblings_root = os.path.dirname(abs_workspace) or abs_workspace

    def _validate_bash(command: str) -> PermissionResult | None:
        """Return a Deny result if *command* is rejected, else ``None``."""
        for pat in deny_patterns:
            if pat.search(command):
                return PermissionResultDeny(
                    message=f"Command matches deny pattern: {pat.pattern}"
                )

        segments = _split_segments(command)
        if not segments:
            return PermissionResultDeny(
                message="Empty command — blocked for safety."
            )

        # Track the effective cwd across chained ``cd`` segments.
        effective_cwd = abs_workspace

        for segment in segments:
            binary = _segment_binary(segment)
            if binary is None:
                return PermissionResultDeny(
                    message=f"Could not parse segment '{segment}' — blocked."
                )
            if binary not in allowlist:
                return PermissionResultDeny(
                    message=(
                        f"Binary '{binary}' is not in the allowlist "
                        f"(segment: '{segment}')."
                    )
                )

            # Path-confinement checks.
            for path in _extract_paths(segment):
                # Only bother with paths that look like filesystem references.
                if not (path.startswith("/") or path.startswith("./")
                        or path.startswith("../") or "/" in path):
                    continue
                candidate = path if os.path.isabs(path) else os.path.join(
                    effective_cwd, path
                )
                # Reject references that land inside another sibling workspace
                # (same parent as our workspace but a different leaf).
                if _is_inside(candidate, siblings_root) and not _is_inside(
                    candidate, abs_workspace
                ):
                    return PermissionResultDeny(
                        message=(
                            f"Path '{path}' targets another workspace sibling "
                            f"outside '{abs_workspace}'."
                        )
                    )

            # If the segment is a ``cd``, update effective_cwd for subsequent
            # segments in the same chain and refuse leaves out of the workspace.
            if binary == "cd":
                tokens = shlex.split(segment)
                # ``cd`` with no arg returns to $HOME — refuse.
                if len(tokens) < 2:
                    return PermissionResultDeny(
                        message="Bare 'cd' is not allowed (would leave workspace)."
                    )
                target = tokens[1]
                new_cwd = target if os.path.isabs(target) else os.path.join(
                    effective_cwd, target
                )
                if not _is_inside(new_cwd, abs_workspace):
                    return PermissionResultDeny(
                        message=(
                            f"'cd {target}' would leave workspace "
                            f"'{abs_workspace}'."
                        )
                    )
                effective_cwd = os.path.realpath(new_cwd)

        return None

    async def can_use_tool(
        tool_name: str,
        input_data: dict,
        context: ToolPermissionContext,
    ) -> PermissionResult:
        # --- Bash guard ---
        if tool_name == "Bash":
            command = input_data.get("command", "")
            deny = _validate_bash(command)
            if deny is not None:
                return deny

        # --- Filesystem confinement for Write / Edit ---
        if tool_name in ("Write", "Edit"):
            file_path = input_data.get("file_path", "")
            if not _is_inside(file_path, abs_workspace):
                return PermissionResultDeny(
                    message=f"Path '{file_path}' is outside workspace '{abs_workspace}'."
                )

        return PermissionResultAllow()

    return can_use_tool


# Type alias for documentation purposes.
CanUseToolCallback = type(create_permission_handler)

File: src/security.ts
/**
 * Module de sécurité — allowlist Bash et confinement filesystem.
 *
 * Fournit une fabrique de callback `can_use_tool` pour le SDK Claude Agent qui
 * impose des allowlists de commandes et un accès fichier limité au workspace.
 *
 * Port fidèle de la V1 Python (`security.py`).
 */

import path from "node:path";

import type { CanUseTool, PermissionResult } from "@anthropic-ai/claude-agent-sdk";

import type { Config } from "./types";

const SEGMENT_SPLIT_RE = /&&|\|\||;|\||\n/;

/**
 * Découpe une commande shell en segments de premier niveau.
 *
 * Découpe sur ``&&``, ``||``, ``;``, ``|`` et les retours à la ligne. Le quoting
 * shell n'est pas pris en compte — un segment coupé à l'intérieur de quotes serait
 * mal détecté mais échoue tout de même en sécurité car chaque moitié est re-validée.
 */
function splitSegments(command: string): string[] {
  return command
    .split(SEGMENT_SPLIT_RE)
    .map((seg) => seg.trim())
    .filter((seg) => seg.length > 0);
}

/**
 * Équivalent minimal de Python ``shlex.split`` : gère les espaces, les quotes
 * simples ``'...'`` et doubles ``"..."``. Renvoie ``[]`` si une quote n'est pas fermée.
 */
function shlexSplit(s: string): string[] {
  const tokens: string[] = [];
  let current = "";
  let hasToken = false;
  let quote: '"' | "'" | null = null;
  let i = 0;

  while (i < s.length) {
    const ch = s[i] as string;

    if (quote !== null) {
      if (ch === quote) {
        quote = null;
      } else {
        current += ch;
      }
      i += 1;
      continue;
    }

    if (ch === '"' || ch === "'") {
      quote = ch;
      hasToken = true;
      i += 1;
      continue;
    }

    if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r" || ch === "\f" || ch === "\v") {
      if (hasToken) {
        tokens.push(current);
        current = "";
        hasToken = false;
      }
      i += 1;
      continue;
    }

    current += ch;
    hasToken = true;
    i += 1;
  }

  if (quote !== null) {
    // Quote non fermée — équivalent du ValueError de shlex.
    return [];
  }

  if (hasToken) {
    tokens.push(current);
  }

  return tokens;
}

/**
 * Extrait le binaire effectif d'un seul segment.
 *
 * Gère :
 *   - Commandes simples : ``npm install``
 *   - Commandes préfixées d'env : ``NODE_ENV=production npm start``
 */
function segmentBinary(segment: string): string | null {
  const tokens = shlexSplit(segment);
  if (tokens.length === 0) {
    return null;
  }
  for (const token of tokens) {
    if (token.includes("=") && !token.startsWith("-")) {
      continue;
    }
    return path.basename(token);
  }
  return null;
}

/** Renvoie les tokens non-flag d'un segment (chemins filesystem candidats). */
function extractPaths(segment: string): string[] {
  const tokens = shlexSplit(segment);
  return tokens.slice(1).filter((t) => !t.startsWith("-") && !t.includes("="));
}

/** Renvoie *true* ssi ``p`` se résout à l'intérieur de ``root``. */
function isInside(p: string, root: string): boolean {
  const real = path.resolve(p);
  return real === root || real.startsWith(root + path.sep);
}

/**
 * Renvoie un callback async ``can_use_tool`` lié à *config* et *workspaceDir*.
 *
 * Le callback impose :
 * 1. **Allowlist de commandes Bash** — chaque segment d'une commande chaînée doit
 *    utiliser un binaire listé dans ``config.security.bash_allowlist``.
 * 2. **Patterns de refus Bash** — les commandes correspondant à un pattern de
 *    ``config.security.bash_denylist`` sont bloquées.
 * 3. **Confinement de chemins Bash** — tout ``cd <path>`` doit cibler un répertoire
 *    à l'intérieur de *workspaceDir*, et aucun segment ne peut référencer un chemin
 *    absolu hors du workspace.
 * 4. **Confinement filesystem pour Write / Edit / Read** — ``file_path`` doit se
 *    résoudre à l'intérieur de *workspaceDir*.
 */
export function createPermissionHandler(config: Config, workspaceDir: string): CanUseTool {
  const allowlist = new Set<string>(config.security.bash_allowlist);
  const denyPatterns: RegExp[] = config.security.bash_denylist.map(
    (p) => new RegExp(escapeRegExp(p)),
  );
  const absWorkspace = path.resolve(workspaceDir);

  function validateBash(command: string): PermissionResult | null {
    for (const pat of denyPatterns) {
      if (pat.test(command)) {
        return {
          behavior: "deny",
          message: `Command matches deny pattern: ${pat.source}`,
        };
      }
    }

    const segments = splitSegments(command);
    if (segments.length === 0) {
      return {
        behavior: "deny",
        message: "Empty command — blocked for safety.",
      };
    }

    // Suit le cwd effectif à travers les segments ``cd`` chaînés.
    let effectiveCwd = absWorkspace;

    for (const segment of segments) {
      const binary = segmentBinary(segment);
      if (binary === null) {
        return {
          behavior: "deny",
          message: `Could not parse segment '${segment}' — blocked.`,
        };
      }
      if (!allowlist.has(binary)) {
        return {
          behavior: "deny",
          message: `Binary '${binary}' is not in the allowlist (segment: '${segment}').`,
        };
      }

      // Vérifications de confinement de chemins — refuse tout argument chemin
      // qui se résout hors du workspace (pas seulement les siblings).
      for (const candidatePath of extractPaths(segment)) {
        // Ne s'intéresse qu'aux chemins qui ressemblent à des références filesystem.
        if (
          !(
            candidatePath === ".." ||
            candidatePath.startsWith("/") ||
            candidatePath.startsWith("./") ||
            candidatePath.startsWith("../") ||
            candidatePath.includes("/")
          )
        ) {
          continue;
        }
        const candidate = path.isAbsolute(candidatePath)
          ? candidatePath
          : path.join(effectiveCwd, candidatePath);
        if (!isInside(candidate, absWorkspace)) {
          return {
            behavior: "deny",
            message: `Path '${candidatePath}' resolves outside workspace '${absWorkspace}'.`,
          };
        }
      }

      // Si le segment est un ``cd``, met à jour effectiveCwd pour les segments
      // suivants de la chaîne et refuse de sortir du workspace.
      if (binary === "cd") {
        const tokens = shlexSplit(segment);
        // ``cd`` sans argument retourne à $HOME — refuse.
        if (tokens.length < 2) {
          return {
            behavior: "deny",
            message: "Bare 'cd' is not allowed (would leave workspace).",
          };
        }
        const target = tokens[1] as string;
        const newCwd = path.isAbsolute(target) ? target : path.join(effectiveCwd, target);
        if (!isInside(newCwd, absWorkspace)) {
          return {
            behavior: "deny",
            message: `'cd ${target}' would leave workspace '${absWorkspace}'.`,
          };
        }
        effectiveCwd = path.resolve(newCwd);
      }
    }

    return null;
  }

  const canUseTool: CanUseTool = async (toolName, input, _options) => {
    // --- Garde Bash ---
    if (toolName === "Bash") {
      const command = input.command === undefined ? "" : String(input.command);
      const deny = validateBash(command);
      if (deny !== null) {
        return deny;
      }
    }

    // --- Confinement filesystem pour Read / Write / Edit ---
    if (toolName === "Read" || toolName === "Write" || toolName === "Edit") {
      const filePath = input.file_path === undefined ? "" : String(input.file_path);
      if (!isInside(filePath, absWorkspace)) {
        return {
          behavior: "deny",
          message: `Path '${filePath}' is outside workspace '${absWorkspace}'.`,
        };
      }
    }

    // --- Confinement filesystem pour Glob / Grep (paramètre path) ---
    if (toolName === "Glob" || toolName === "Grep") {
      const searchPath = input.path === undefined ? "" : String(input.path);
      if (searchPath && !isInside(searchPath, absWorkspace)) {
        return {
          behavior: "deny",
          message: `Search path '${searchPath}' is outside workspace '${absWorkspace}'.`,
        };
      }
    }

    return { behavior: "allow" };
  };

  return canUseTool;
}

/** Échappe les métacaractères regex (équivalent de Python ``re.escape``). */
function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
